### PR TITLE
Refactor BRB Login by separating the login view from the accounts view

### PR DESCRIPTION
**Pull request by whoiswillma**
Issue: #202 

- Separate the login view from the accounts view by making the login view into a view controller, and adding an additional view controller `BRBViewController` to manage the `BRBLoginViewController` and `BRBAccountViewController`. 
  * This was previously accomplished by adjusting the view alpha on each view. 
- Login View Changes:
   * Add new error view
   * Use a `UISwitch` to save login info

Simulator Screen Shot - iPhone 8 - 2019-09-02 at 14 06 43

### CODE REVIEW

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: delete if unused

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: I believe doing `edges` already includes `top` so you prob just need `make.edges.equalToSuperview()`

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: Delete if unused or add some sort of description

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift:58]: Maybe add a comment for what the 4 sections are.

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: Can we make the cell identifiers a constant since we use it in more than one place?

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: nit: `"$(account.brbs)"`

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: same here and for `case 2:` as well

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: you can get rid of `else` since you are returning on line 84

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: same here

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBLoginViewController.swift]: Add description or delete

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBLoginViewController.swift]: remove the `else` since you have `return` on line 46

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBLoginViewController.swift:61]: can we make this private?

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBLoginViewController.swift]: delete or add description

File: Eatery/Controllers/BRB/BRBPrivacyStatementViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBPrivacyStatementViewController.swift]: delete or add description

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBViewController.swift]: delete or add description

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBViewController.swift:113]: delete whitespace

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBViewController.swift]: delete or add description

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBViewController.swift:145]: delete whitespace

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBViewController.swift]: delete or add description

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBViewController.swift:158]: delete whitespace

File: Eatery/Views/BRB/BRBLoginErrorView.swift
Comment (kevinchan159) [inline Eatery/Views/BRB/BRBLoginErrorView.swift]: delete or add description

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (chalo2000) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: Is this intentional or did you forget to put something here?

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (chalo2000) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]:
```suggestion
            let history = account.history[indexPath.row]
```

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (chalo2000) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: ditto

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (chalo2000) [inline Eatery/Controllers/BRB/BRBLoginViewController.swift]: ditto

**Review by kevinchan159 — CHANGES_REQUESTED** [review]
awesome work!!! 😍 great job following the PR guidelines and adding a screenshot! 
just a small note, it might be better to make the PR's a bit smaller in the future just so that its easier for people to look at past commits to help debug any issues that may arise!

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (chalo2000) [inline Eatery/Controllers/BRB/BRBLoginViewController.swift]: ditto

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (chalo2000) [inline Eatery/Controllers/BRB/BRBViewController.swift]: ditto

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (chalo2000) [inline Eatery/Controllers/BRB/BRBViewController.swift]: ditto

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (chalo2000) [inline Eatery/Controllers/BRB/BRBViewController.swift]: ditto

**Review by chalo2000 — CHANGES_REQUESTED** [review]
Other than minuscule issues, LGTM

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (Omarrasheed) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: Can we move this setup stuff into a cell.configure().

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/BRB/BRBAccountViewController.swift]: This was just a habit I got into over the summer. I'll get rid of them to follow the AppDev style guide

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/BRB/BRBLoginViewController.swift:61]: Do you mean make the delegate private? We need write access to it from the `BRBViewController`

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/BRB/BRBViewController.swift:113]: In general should we keep whitespace only lines?

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (Omarrasheed) [reply Eatery/Controllers/BRB/BRBAccountViewController.swift]: You can just store an array of those `titleLabel.text` and `subtitleLabel.text` values at the top and instead of having this switch statement, just directly index that array. That'll make it 10x easier to change the order or make any big structural changes to this stuff down the line.

Could also move the two text params into some struct to ensure they're being paired correctly. This may be overkill tho cc @kevinchan159

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (Omarrasheed) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: nit: delete newline

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (kevinchan159) [reply Eatery/Controllers/BRB/BRBLoginViewController.swift:61]: oh i see. could we pass pass in the `delegate` by creating a custom init for `BRBLoginViewController` and then change the `delegate` to be private? This way, the delegate will be set only once and we prevent the possibility of someone changing it later. We should make all delegates private because other classes shouldn't be accessing them.

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (kevinchan159) [reply Eatery/Controllers/BRB/BRBViewController.swift:113]: its fine to have newlines if it helps with reading code but i like to delete unnecessary whitespace like in this case

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (Omarrasheed) [inline Eatery/Controllers/BRB/BRBLoginViewController.swift]: can we unwrap self with a guard let?

File: Eatery/Views/BRB/BRBBalanceTableViewCell.swift
Comment (Omarrasheed) [inline Eatery/Views/BRB/BRBBalanceTableViewCell.swift]: This is cool i didn't know this was a thing. Def gonna use this

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/BRB/BRBViewController.swift:113]: What about whitespace between lines of code that are not top-level?

```
class A {
    func x() {
        ...
    }
____< whitespace here?
    func y() {
        ...
    }
}
```

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [reply Eatery/Controllers/BRB/BRBAccountViewController.swift]: i think the struct is a bit of an overkill

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (kevinchan159) [reply Eatery/Controllers/BRB/BRBViewController.swift:113]: i would also prefer to not have whitespace there

Comment (whoiswillma) [reply Eatery/Controllers/BRB/BRBViewController.swift:113]: Ok, I've enabled the "Automatically Trim Whitespace only lines" feature in Xcode. There are tons of lines that are whitespace only that I've committed (many many more than are mentioned in PR comments). Instead of going back through and deleting them all individually I'd rather just let them dissipate over time as the files get edited.

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (young-k) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift:6]: Could we update this to be `Cornell AppDev`, cc @kevinchan159

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (young-k) [inline Eatery/Controllers/BRB/BRBLoginViewController.swift:85]: probably cleaner to have a newline before this one

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/BRB/BRBLoginViewController.swift:61]: Seems like quite a bit of overhead for a pattern that's well established: The only owner of the `BRBLoginViewController`(the `BRBViewController`), has its reference to it private, so there's no chance of it being changed unintentionally outside of the owners file. 

Delegation is a well established pattern within Apple's frameworks, and it (almost) always works by assigning to a weak variable. Having that consistency with Apple's frameworks is also important. 

It's true that other classes shouldn't be accessing delegates, but I think that it's generally understood that delegates should only be assigned to. It seems that someone would have to be intentionally writing malicious code or just be uninformed to access the delegate in that way.

File: Eatery/Controllers/BRB/BRBViewController.swift
Comment (Omarrasheed) [reply Eatery/Controllers/BRB/BRBViewController.swift:113]: swiftlint autocorrect should get all of em for u

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (whoiswillma) [reply Eatery/Controllers/BRB/BRBLoginViewController.swift]: I think the better action here would just be to retain a strong reference to `self`. In the first case, the `UIViewPropertyAnimator` immediately falls out of scope, and in the second case the closure never even leaves the function scope

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: can we add a `func configure(with title: string, subtitle:string) {}` function to the cell and just pass these values in there instead?

File: Eatery/Controllers/BRB/BRBAccountViewController.swift
Comment (kevinchan159) [inline Eatery/Controllers/BRB/BRBAccountViewController.swift]: `"-$\(history.amount)"`

File: Eatery/Controllers/BRB/BRBLoginViewController.swift
Comment (kevinchan159) [reply Eatery/Controllers/BRB/BRBLoginViewController.swift:61]: Oh I think there was a bit of confusion. I totally agree that the delegate should be `weak`. I was just saying I think setting it to be `private` in addition would help. I also agree that in 99% of the cases, people probably won't write malicious code but I think additional layers of security are always good. The argument is similar for why we set view variables as private. It is understood that we shouldn't be setting these view variables outside of the class but making them `private` just helps to guarantee it. Passing the delegate through the init also guarantees that we only set the delegate once.

**Review by kevinchan159 — APPROVED** [review]
approving to unblock but just left 3 more comments!

**Review by chalo2000 — APPROVED** [review]

**Pull request merged**
